zerver/migrations/0453_backfill_remote_realmauditlog_extradata_to_json_field.py

```diff
@@ -0,0 +1,86 @@
+from django.db.migrations.state import StateApps
+from django.db.models import JSONField, Model
+from django.db.models.functions import Cast, JSONObject
+
+# This migration is mostly the same as realmauditlog_extra_data_json
+
+OLD_VALUE = "1"
+NEW_VALUE = "2"
+USER_FULL_NAME_CHANGED = 124
+BATCH_SIZE = 5000
+
+
+def do_bulk_backfill_extra_data(
+    audit_log_model: Type[Model], id_lower_bound: int, id_upper_bound: int
+) -> None:
+    # First handle the special case for audit logs with the USER_FULL_NAME_CHANGED event, which stores
+    # the full name not as str(dict()) but a plain str.
+    # Note that we only update the entries where extra_data_json has the default value,
+    # because we do not want to override existing audit log entries with a NEW_VALUE of None for extra_data_json.
+    # We do not need to skip existing entries for other parts of backfilling because we have double-write implemented
+    # so that the backfilled value will still be consistent.
+    audit_log_model.objects.filter(
+        event_type=USER_FULL_NAME_CHANGED,
+        id__range=(id_lower_bound, id_upper_bound),
+        extra_data_json={},
+    ).update(extra_data_json=JSONObject(**{OLD_VALUE: "extra_data", NEW_VALUE: None}))
+
+    # A dict converted with str() will start with a open bracket followed by a single quote,
+    # as opposed to a JSON-encoded value, which will use a _double_ quote.
+    # We use this to filter out those entries with malformed extra_data to be handled later.
+    # This should only update rows with extra_data populated with orjson.dumps.
+    (
+        audit_log_model.objects.filter(
+            extra_data__isnull=False, id__range=(id_lower_bound, id_upper_bound)
+        )
+        .exclude(extra_data__startswith="{'")
+        .exclude(event_type=USER_FULL_NAME_CHANGED)
+        .update(extra_data_json=Cast("extra_data", output_field=JSONField()))
+    )
+
+    python_valued_audit_log_entries = audit_log_model.objects.filter(
+        extra_data__startswith="{'", id__range=(id_lower_bound, id_upper_bound)
+    )
+    for audit_log_entry in python_valued_audit_log_entries:
+        # extra_data for entries that store dict stringified with builtins.str() are converted
+        # back with ast.literal_eval for safety and efficiency.
+        # str()'d extra_data with the REALM_DISCOUNT_CHANGED event type is not handled by this
+        # migration. We expect that all such entries are manually converted beforehand or an
+        # error will occur during the migration, because ast.literal_eval does not allow the
+        # evaluation of Decimal.
+        audit_log_entry.extra_data_json = ast.literal_eval(audit_log_entry.extra_data)  # type: ignore[attr-defined] # The migration cannot depend on zerver.models, which contains the real type of the RealmAuditLog model, so it cannot be properly typed.
+    audit_log_model.objects.bulk_update(python_valued_audit_log_entries, fields=["extra_data_json"])
```

Comment:
@alexmv I think we should plan for you to test versions of this migration on Zulip Cloud, which should have all our weird legacy/historical shapes for these events, before we run it in Zulip Cloud. I guess we can schedule it after #25878 makes it to production, or run variants of these commands to just do the select version, but the latter seems kinda annoying.